### PR TITLE
#166185762 Rest Api Registration

### PR DESCRIPTION
**Pull request by NicholusMuwonge**
#### What does this PR do?
allows third-party applications to create users using the rest API . This is through making ```/api/v2/user/creation/```,```add-user-rest-api``` **and** ```list-user-rest-api```  functional.

#### Description of Task to be completed?
An application that is signed up as an ordinary user can create different users from its website using the wger custom commands that run in the terminal. These apps require a special key to be able to create other users.

#### How should this be manually tested?
- one should clone this repository and install the required dependencies
- log in to your local server supported front end and get a special key to allow you to create other users. It should look like this;

Screen Shot 2019-06-03 at 17 47 52

-  head on over to your terminal impersonating the third party application. start the local server in another terminal and populate the other terminal like:
``` python manage.py add-user-rest-api <special-key-from-website> <app-username>```
you should be able to see a green line of text stating 
```granted permission to create users```

At this point, one ought to run over to postman and test this endpoint for creating a user with the setup as demonstrated below on success:
Screen Shot 2019-06-06 at 14 31 20

-  check a list of users created by that particular person or to check who created certain users, just keep server running and populate terminal with:
```<list-user-rest-api> <app-username> ``` and a fancy blue table will be output with two columns, one being for creator and username. Something like
Screen Shot 2019-06-06 at 14 33 28

#### Any background context you want to provide?
I got some of the ideas from past wger repo projects.
I also read [here](https://labofcoding.com/custom-django-management-commands/)
I also altered the Circle Ci badge to fix the current bug with this PR.
When merged, one will be able to use Heroku Terminal to run these commands

#### What are the relevant pivotal tracker stories?
[#166185762](https://www.pivotaltracker.com/story/show/166185762)

### CODE REVIEW

**Review by richien — COMMENTED** [review]
Hello @NicholusMuwonge, this is good work though I am struggling to understand the implementation. According to the PT description, `REST API consumers that want to create users` and `create new accounts over the API` seem to be the distinctive features, (keywords being, **create users via the REST API**). I am wondering where the **endpoints** are that third-party apps will use to create the users? Kindly note the Heroku command line is meant for internal management of our application and not for third parties to query our application. For instance, how will third-party apps login to our app's Heroku instance to run `add-user-rest-api` as you have mentioned in your PR? Kindly clarify this implementation.

**Comment by NicholusMuwonge** [comment]
> Hello @NicholusMuwonge, this is good work though I am struggling to understand the implementation. According to the PT description, `REST API consumers that want to create users` and `create new accounts over the API` seem to be the distinctive features, (keywords being, **create users via the REST API**). I am wondering where the **endpoints** are that third-party apps will use to create the users? Kindly note the Heroku command line is meant for internal management of our application and not for third parties to query our application. For instance, how will third-party apps login to our app's Heroku instance to run `add-user-rest-api` as you have mentioned in your PR? Kindly clarify this implementation.

Now, for the story [#166185762](https://www.pivotaltracker.com/story/show/166185762), it would require one to be able to grant access for external applications that are signed up as users to create other users, now there are two ways I could go about this implementation being, the UI way that will then require endpoints and according to my understanding, a third party app will automate the creation of users hence no need for a UI. Which Is why I go with the 2nd option that was also deemed the easier way on the story. The implementation is that on completion of this and the application, one who wants to create users will install Wger (not clone it) and these commands will come in the package like ```wger create_settings``` and every time this command will be run, new users will be created.

**Comment by richien** [comment]
> > Hello @NicholusMuwonge, this is good work though I am struggling to understand the implementation. According to the PT description, `REST API consumers that want to create users` and `create new accounts over the API` seem to be the distinctive features, (keywords being, **create users via the REST API**). I am wondering where the **endpoints** are that third-party apps will use to create the users? Kindly note the Heroku command line is meant for internal management of our application and not for third parties to query our application. For instance, how will third-party apps login to our app's Heroku instance to run `add-user-rest-api` as you have mentioned in your PR? Kindly clarify this implementation.
> 
> Now, for the story [#166185762](https://www.pivotaltracker.com/story/show/166185762), it would require one to be able to grant access for external applications that are signed up as users to create other users, now there are two ways I could go about this implementation being, the UI way that will then require endpoints and according to my understanding, a third party app will automate the creation of users hence no need for a UI. Which Is why I go with the 2nd option that was also deemed the easier way on the story. The implementation is that on completion of this and the application, one who wants to create users will install Wger (not clone it) and these commands will come in the package like `wger create_settings` and every time this command will be run, new users will be created.

Thank you for the clarification, though it raises some more questions. When I re-read the story [#166185762](https://www.pivotaltracker.com/story/show/166185762), I take two things away from it. 1. that there needs to be an API route for creating Users and 2. That there needs to be a management interface for granting permission to API Users to create other users. The story seems to suggest that the **management interface** can be either a UI or a Command line interface whose purpose is to grant an API user, permission to create other Users via the API endpoint that you would create in 1. In other words, the expectation I had was that an admin user of the 'wger' app would grant permission to create users to an API consumer using the `add-user-rest-api` interface; which would then enable the User to register Users with `wger` via the API endpoint. Based on the explanation you have given above, I get the impression that third-party applications would have to run there own instances of `wger`, in which case they could as well directly create users without the need for the API since they would be the admins of the application. I believe the story implies that it is easier to create a command line interface for granting permissions rather than a UI. Perhaps my understanding of the story is flawed, based on your explanation, but if the story speaks of registering a User via a **REST API**, I was expecting to test an endpoint.

File: wger/core/api/views.py [hunk 1]
```diff
@@ -125,3 +137,56 @@ class WeightUnitViewSet(viewsets.ReadOnlyModelViewSet):
...
+        if not permission:
+            return Response(
+                'You arent authorised to create user contact admin for rights',
+                status=status.HTTP_400_BAD_REQUEST)
+        serializer = self.serializer_class(data=request.data)
+        serializer = self.serializer_class(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        permission = UserProfile.objects.filter(
+            user=self.request.user.id, is_allowed=True)
+        if User.objects.filter(
```
Comment (nakatuddesuzan) [inline wger/core/api/views.py:160]: Great work @NicholusMuwonge. you have good variable namings. I have a concern here, you can do a little bit of refactoring in this class and put some conditions in a separate file which can be like a helpers file. Then import them and use them in this class. With this, we shall be able to improve our code concerning code climate issues.

File: wger/core/management/commands/list-user-rest-api.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+from django.core.management.base import (
+    BaseCommand, CommandError)
+from wger.core.models import (
+    ApiUserModel, User)
+from django_tabulate import tabulate_qs
+from tabulate import tabulate
+
+
+class Command(BaseCommand):
```
Comment (nakatuddesuzan) [inline wger/core/management/commands/list-user-rest-api.py:9]: Kindly use a more descriptive name to show what these commands are for. Say `CreateUserCommand`

File: wger/core/api/views.py [hunk 3]
```diff
@@ -125,3 +137,56 @@ class WeightUnitViewSet(viewsets.ReadOnlyModelViewSet):
...
+            config.user = user
+            config.save()
+        user.userprofile.save()
+        # update the APIUser table
+
+        new_user = ApiUserModel.objects.create(
+            user_being_created=user,
+            creator=User.objects.get(pk=user_id_check.user_id))
+        new_user.save()
+        return Response(serializer.data, status=status.HTTP_201_CREATED)
```
Comment (nakatuddesuzan) [inline wger/core/api/views.py:192]: This class has very many lines of code. Consider the refactoring suggested above

File: wger/core/api/views.py (hunk 1 above)
Comment (NicholusMuwonge) [reply wger/core/api/views.py:160]: okay, let me improve on that

**Comment by nakatuddesuzan** [comment]
Also, look into why the build is failing and try to fix the issue

File: wger/core/management/commands/list-user-rest-api.py (hunk 2 above)
Comment (NicholusMuwonge) [reply wger/core/management/commands/list-user-rest-api.py:9]: It's more of a must according to Django docs that that should be Command

Comment (nakatuddesuzan) [reply wger/core/management/commands/list-user-rest-api.py:9]: Okay noted

**Pull request merged**